math/smatrix/inc/Math/CholeskyDecomp.h

```diff
@@ -312,14 +323,43 @@ template<class F> class CholeskyDecompGenDim
 private:
    /** @brief dimensionality
     * dimensionality of the problem */
-   unsigned fN;
+   unsigned fN = 0;
    /// lower triangular matrix L
    /** lower triangular matrix L, packed storage, with diagonal
     * elements pre-inverted */
-   F *fL;
+   std::unique_ptr<F[]> fL;
```

Comment:
```suggestion
   std::vector<F> fL;
```
The usual C++ class for automatic memory management of dynamic arrays is `std::vector`. Why not use it here?

Reply:
In principle, I agree. But since we fill the array as we progress in the computation, with std::vector, we'd either have to pay for filling an array of Fs which we'll override immediately, or reserve, and then pay for the check if the array is large enough for every number we will. You may argue that the expensive operation is likely going to be the memory allocation itself (which is likely true), but once the memory is allocated, this version does less work.
Do you still want this changed to std::vector?

Reply:
It occurs to me that if we do go down the std::vector path, we should offer a way to "use a stolen workspace", so that if you do many decompositions, you can get away reusing the same std::vector over and over, moving it from decomposition object to decomposition object, so you pay for the memory allocation only once. I have no good idea how the interface for something like that might look like, though...

Reply:
Hi @manuelschiller thanks for responding here and initiating the discussion!

It's true that `std::vector` has theoretically some overhead because of the zero-initialization of the array, but as you say it's negligible compared to the allocation time and it's such a common thing to do that the compiler can optimize it very well. I experimented a bit with it:

```c++
// Compile with g++ -O(0,1,2, or 3) -o testVector testVector.cxx

#include <chrono>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class TimingScope {
public:
   TimingScope(std::string const &label) : _begin{std::chrono::steady_clock::now()}, _label{label} {}
   ~TimingScope()
   {
      std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
      std::cout << "Time difference for " << _label << " = "
                << std::chrono::duration_cast<std::chrono::microseconds>(end - _begin).count() << " [µs]" << std::endl;
   }

private:
   std::chrono::steady_clock::time_point _begin;
   const std::string _label;
};

using F_t = float;

// Do something with the array, including printout, so that the optimizer
// doesn't optimize everything away.
void doSomething(F_t const* arr, std::size_t n)
{
   F_t sum = 0.0;
   for (std::size_t i = 0; i < n; ++i) {
      sum += arr[i];
   }
   std::cout << sum << std::endl;
}

int main()
{
   std::size_t n = 1000000;

   {
      TimingScope ts{"std::unique_ptr<F_t[]>"};
      auto vec1 = std::make_unique<F_t[]>(n);
      doSomething(vec1.get(), n);
   }
   {
      TimingScope ts{"std::vector<F_t>"};
      std::vector<F_t> vec2(n);
      doSomething(vec2.data(), n);
   }
}
```

Using gcc, the `std::vector` creation is already consistently faster on my platform starting with `O1` optimizations. Also, including the extra `<vector>` header should not be controversial because the STL is available everywhere. Actually it's not an extra header, because then you don't need `<memory>`.

If you insist on using the `unique_ptr` we can do it, but given that `std::vector` is a) the standard thing to do, b) it was no significant performance overhead and even appears to get optimize better, and c) it greatly simplifies the code reducing the margin for error, I would still like the change to `std::vector`

Reply:
Well, good point, and I have not done the timing check yet (which I will do using callgrind, I think). I have also implemented both options (although I compile at -O2, because CholeskyDecompGenDim is meant to be used in extremely tight code that gets the hell optimized out of it), and I will do the timing check on that. That said, for my test program which fits a parabola from 3 points, I tried both the std::vector and the std::unique_ptr version, and the std::unique_ptr version is around 500 bytes of code shorter (for a 10k test program), so vector does have costs.

Well, I'll try to get a timing measurement that I trust (i.e. not affected by the OS preempting like in the code above, and with a real CholeskyDecompGenDim workload because optimizers are scarily clever these days, and you often have to look at the assembly to make sure that things did not get optimized away completely), but it may be a day or two before I can complete this because I've got a 9 month old daughter at home who's running a fever...